README
NAME
    Class::Data::Inheritable::Translucent - Inheritable, overridable,
    translucent class data / object attributes

SYNOPSIS
      package Foo;
      use parent qw(Class::Data::Inheritable::Translucent);
      sub new { bless {}, shift }

      # Read/write attributes
      Foo->mk_class_accessor(cattr => "bar");
      Foo->mk_translucent_accessor(tattr => "bar");
      Foo->mk_object_accessor(oattr => "bar");

      # Read-only attributes
      Foo->mk_ro_class_accessor(foo => 1);
      Foo->mk_ro_translucent_accessor(bar => 1);
      Foo->mk_ro_object_accessor(baz => 1);

      my $obj = Foo->new;
      print $obj->cattr; # prints "bar"
      print $obj->tattr; # prints "bar"
      print $obj->oattr; # prints "bar"
      print Foo->cattr;  # prints "bar"
      print Foo->tattr;  # prints "bar"

      $obj->cattr("baz");
      $obj->tattr("baz");
      $obj->oattr("baz");
      print $obj->cattr; # prints "baz"
      print $obj->tattr; # prints "baz"
      print $obj->oattr; # prints "baz"
      print Foo->cattr;  # prints "baz"
      print Foo->tattr;  # prints "bar"

      Foo->cattr("qux");
      Foo->tattr("qux");
      print $obj->cattr; # prints "qux"
      print $obj->tattr; # prints "baz"
      print Foo->cattr;  # prints "qux"
      print Foo->tattr;  # prints "qux"

      delete $obj->{tattr};
      delete $obj->{oattr};
      print $obj->tattr; # prints "qux"
      print $obj->oattr; # prints "bar"
      print Foo->tattr;  # prints "qux"

      package Bar;
      use parent qw(Class::Data::Inheritable::Translucent::Object);

      Bar->mk_object_accessor(attr => 0);

      # Class constructor
      my $obj1 = Bar->new();
      my $obj2 = Bar->new(attr => 1);
      my $obj3 = Bar->new($obj2);
      my $obj4 = Bar->new($obj2, attr => 2);

      # Object clone
      my $obj5 = $obj2->clone();
      my $obj6 = $obj2->clone(attr => 3);

      # Object copy
      my $obj7 = $obj2->copy();

      # Object reset
      $obj7->reset();

      package Bar2;
      use parent qw(Bar);

      my $_Count = 0;

      # Separated, overridable, object initialization
      sub initialize {
          my($self, %args) = @_;
          $_Count++;
          return $self->SUPER::initialize(%args);
      }

      sub DESTROY {
          $_Count--;
      }

      package Baz;
      use parent qw(Class::Data::Inheritable::Translucent::Singleton);

      Baz->mk_object_accessor(attr => 0);

      # Class constructor
      my $obj1 = Baz->instance(attr => 1);
      my $obj2 = Baz->instance(attr => 2); # same object as $obj1

      # Object reset
      $obj1->reset(); # $obj1 and $obj2 now both have attr == 0

      package Baz2;
      use parent qw(Baz);

      my $_Count = 0;

      # Separated, overridable, object initialization
      sub initialize {
          my($self, %args) = @_;
          $_Count++;
          return $self->SUPER::initialize(%args);
      }

      sub DESTROY {
          $_Count--;
      }

DESCRIPTION
    This module is based on Class::Data::Inheritable; the
    mk_class_accessor() method is largely the same as
    Class::Data::Inheritable's mk_classdata() method, but
    mk_translucent_accessor() creates class data accessors that double as
    translucent object attribute accessors. Accessors for plain object
    attributes can also be created with mk_object_accessor(). Read-only
    accessors for all three types of data/attributes can also be created
    with mk_ro_class_accessor(), mk_ro_translucent_accessor() and
    mk_ro_object_accessor().

    Your class should inherit from Class::Data::Inheritable::Translucent and
    call the relevant mk_*_accessor() methods during its initialization. If
    you are making use of object accessor methods (translucent or otherwise)
    then your class will also need to have a constructor method to create
    objects. By default, the objects are assumed to be hash references with
    their attributes stored directly inside them, i.e. the value of object
    $obj's attribute $attr is assumed to be stored in $obj->{$attr}. See the
    attrs() method on how to change that. If a simple hash-based object is
    all you need then you can have a constructor method provided for you by
    inheriting from Class::Data::Inheritable::Translucent::Object or
    Class::Data::Inheritable::Translucent::Singleton instead.

    Class::Data::Inheritable::Translucent::Object is an abstract base class
    extending Class::Data::Inheritable::Translucent with constructor/clone
    methods with separated initialization. The objects constructed are
    simple hash-based objects compatible with the accessor methods installed
    by Class::Data::Inheritable::Translucent.

    Attributes may be initialized by the constructor, either explicitly, or
    by copying another object, or via a separated initialize() method which
    can be overridden in subclasses. A copy() method is also provided to
    copy the attributes of one object to another, and a reset() method can
    be used to restore an object's non-read-only attributes to their default
    values (translucent or otherwise).

    Class::Data::Inheritable::Translucent::Singleton is an abstract base
    class extending Class::Data::Inheritable::Translucent with a singleton
    constructor method with separated initialization. The objects
    constructed are simple hash-based objects compatible with the accessor
    methods installed by Class::Data::Inheritable::Translucent.

    The singleton design pattern behaviour is inherited from either
    Apache::Singleton or Class::Singleton, whichever is found first (looking
    in that order); an exception will be thrown if neither is available.
    Apache::Singleton is more flexible since it allows one instance per
    request when run in a mod_perl environment (but still permits only one
    instance per process otherwise).

    Attributes may be initialized by the constructor, either explicitly, or
    via a separated initialize() method. A reset() method can be used to
    restore the object's non-read-only attributes to their default values
    (translucent or otherwise).

    Note that all of these methods only perform shallow copies of attribute
    values. Therefore, different classes and/or objects will end up sharing
    some data if any attribute values are set to references.

COMPATIBILITY
    Before version 2.00 of this module, an object attribute that had been
    set to override translucent class data could be "reset" to reveal the
    inherited class data value by setting it to the undefined value.

    As of version 2.00 of this module, the attribute must now be deleted
    from the object in order to have the same effect. If your objects are
    hashrefs then this is done simply with "delete $object->{$attribute}";
    more generally, and in particular if you have overridden attrs(), this
    is done with "delete $object->attrs()->{$attribute}".

    THIS IS AN INCOMPATIBLE CHANGE. EXISTING SOFTWARE THAT USES THIS FEATURE
    WILL NEED TO BE MODIFIED.

INSTALLATION
    See the INSTALL file.

COPYRIGHT
    Version 0.01 Copyright (C) 2005 Ryan McGuigan. All rights reserved.

    Changes in Version 1.00 onwards Copyright (C) 2009, 2011, 2014-2015
    Steve Hay. All rights reserved.

LICENCE
    This distribution is free software; you can redistribute it and/or
    modify it under the same terms as Perl itself, i.e. under the terms of
    either the GNU General Public License or the Artistic License, as
    specified in the LICENCE file.


Changes
Revision history for Perl distribution Class-Data-Inheritable-Translucent

2.00 Development

    - Added Class::Data::Inheritable::Translucent::Singleton, an abstract base
      class extending Class::Data::Inheritable::Translucent with a singleton
      constructor method with separated initialization. This is a variation of
      Class::Data::Inheritable::Translucent::Object implementing the singleton
      design pattern.

    - Added Class::Data::Inheritable::Translucent::Object, an abstract base
      class extending Class::Data::Inheritable::Translucent with a
      constructor/clone method with separated initialization.  Users of the main
      module will always need a constuctor for their objects and may not already
      have one so it is convenient for this distribution to provide a class that
      fulfils this common need.

    - Fixed broken link for active bugs on the CPAN Request Tracker.

    - Added mk_object_accessor(), which creates a non-translucent object
      attribute accessor.  In this case there is no inherited translucent class
      data for the object to override, but a default value can still be
      specified, and in theory restored later in a similar manner to removing an
      object's override of translucent class data to reveal the inherited class
      data value.  However, it would be quite unnatural to set such an attribute
      to the undefined value in order to restore the original default value, so
      a new, more obvious, behaviour has been introduced:  The attribute must be
      deleted from the object, i.e. delete()d from the attrs() hash.

      The same behaviour has been extended to include translucent object
      attributes as well, partly for consistency, and partly because it is a
      more natural behaviour (and no longer precludes the possibility of
      actually setting an attribute to the undefined value if required!).  Two
      related CPAN modules (Class::Data::Accessor and Class::Accessor::Grouped)
      also behave this way.  This feature was not documented other than being
      illustrated in the SYNOPSIS but because of the change in behaviour that
      some users may have been relying on, the $VERSION has been given a major
      version bump and appropriate notices have been added to the man page and
      README.

    - Renamed mk_translucent() to mk_translucent_accessor() (retaining the old
      name for backwards compatibility) and added mk_class_accessor(), which
      does exactly what mk_classdata() does in Class::Data::Inheritable except
      that it makes use of Sub::Name, supports the use of non hash-based objects
      and doesn't blindly overwrite any existing (accessor) methods.  Thus, if,
      like me, you're bugged that the maintainer of Class::Data::Inheritable has
      never seen fit to address CPAN RT#51228 or CPAN RT#6281, you can now use
      mk_class_accessor() here instead (without your class data accessor taking
      on the double life of a translucent object attribute accessor, as would be
      the case if you simply used mk_translucent_accessor() instead).

    - Made mk_translucent() croak() if wrongly called as an object method.

    - Added optional Changes testing (skipped unless AUTHOR_TESTING).

    - Reformatted Changes file as per CPAN::Changes::Spec.

    - Made Perl Critic, Pod and Pod Coverage tests skip unless AUTHOR_TESTING.

    - Added optional Perl::Critic testing.

    - Made code Perl::Critic clean.

    - Noted other modules that do almost exactly the same thing and documented
      perhaps the biggest difference from them: that accessor methods are not
      installed if the subroutines already exist.  [CPAN RT#96066]

    - Simplified Makefile.PL requirements.

1.04 2014-05-30

    - The modules listed as recommendations of optional features are actually
      requirements of those features.

1.03 2014-05-29

    - Uploaded source code to GitHub and included repository URLs in metadata.

    - Included META.json file in addition to META.yml.

    - Set minimum required ExtUtils::MakeMaker version to 6.64 to ensure that
      all parameters used are supported, to save jumping through hoops to
      support earlier versions.  (This should not be a problem since
      ExtUtils::MakeMaker 6.64 is easily installed into Perl 5.8.1 and above,
      that being the whole point of the new choice of minimum supported Perl
      version.)

    - Set minimum required Perl version to 5.8.1.  This is in line with the
      minimum requirement of the "Perl Toolchain".

1.02 2011-06-03

    - Fixed the default implementation of attrs() to work as documented, so that
      each attribute really is now stored in $obj rather than $obj->{attrs} by
      default.

1.01 2009-11-09

    - Incorporated two improvements based on code in Class::Accessor:

      * Generated accessor methods are now named using Sub::Name if that module
        is available.

      * Existing accessor methods (or their aliases) are no longer overwritten
        by mk_translucent().

1.00 2009-11-05

    - Changed mk_translucent() so that when overriding the translucent class
      default value on a sub-class object it doesn't affect the inherited
      translucent class default value still used by other instances of that
      sub-class.  In fact, it was wrong even to install an override unless the
      attribute is actually being overriddeen on the sub-class itself rather
      than on an instance of the sub-class.  [CPAN RT#51097]

    - Changed Makefile.PL to check $ExtUtils::MakeMaker::VERSION before setting
      ABSTRACT_FROM and AUTHOR.

    - Put the NAME content all on one line in the .pm file so that the
      'abstract' is not truncated in the META.yml file.

    - Added LICENSE to Makefile.PL.

    - Removed redundant 'clean' from Makefile.PL so that it doesn't remove the
      tarball when running the 'clean' target after the 'dist' target.


lib/Class/Data/Inheritable/Translucent/Base.pm
#===============================================================================
#
# lib/Class/Data/Inheritable/Translucent/Base.pm
#
# DESCRIPTION
#   Abstract base class extending Class::Data::Inheritable::Translucent with
#   common functionality (initialize/reset methods) required by its Object and
#   Singleton subclasses.
#
# COPYRIGHT
#   Copyright (C) 2015 Steve Hay.  All rights reserved.
#
# LICENCE
#   This module is free software; you can redistribute it and/or modify it under
#   the same terms as Perl itself, i.e. under the terms of either the GNU
#   General Public License or the Artistic License, as specified in the LICENCE
#   file.
#
#===============================================================================

package Class::Data::Inheritable::Translucent::Base;

use 5.008001;

use strict;
use warnings;

use parent qw(Class::Data::Inheritable::Translucent);

use Carp qw(carp croak);

#===============================================================================
# CLASS INITIALIZATION
#===============================================================================

our($VERSION);

BEGIN {
    $VERSION = '2.00';
}

__PACKAGE__->mk_object_accessor(__CDITB_initialized => 0);

#===============================================================================
# PUBLIC METHODS
#===============================================================================

#-------------------------------------------------------------------------------
# Object methods
#-------------------------------------------------------------------------------

#
# Overridable.
#
sub initialize {
    my($self, %args) = @_;
    croak("initialize() is an object method, not a class method") unless ref $self;

    # If the object has already been initialized (e.g. when called from reset())
    # then we must only (re-)initialize non-read-only attributes.
    my @attrs = $self->_get_editable_attrs();

    foreach my $attr (@attrs) {
        $self->$attr(delete $args{$attr}) if exists $args{$attr};
    }

    if (%args) {
        # Get a list of all attributes to distinguish read-only ones from
        # unknown ones.
        $self->_initializing(1);
        my %all_attrs = map { $_ => 1} $self->_get_editable_attrs();
        $self->_initializing(0);

        my @readonly_attrs;
        my @unknown_attrs;
        foreach my $attr (sort keys %args) {
            if (exists $all_attrs{$attr}) {
                push @readonly_attrs, $attr;
            }
            else {
                push @unknown_attrs, $attr;
            }
        }

        if (@readonly_attrs) {
            carp("initialize() ignored the following read-only attributes: " .
                 join(', ', map { "'$_'" } @readonly_attrs));
        }

        if (@unknown_attrs) {
            carp("initialize() ignored the following unknown attributes: " .
                 join(', ', map { "'$_'" } @unknown_attrs));
        }

        return 0;
    }

    return 1;
}

sub reset {
    my($self) = @_;
    croak("reset() is an object method, not a class method") unless ref $self;

    # Delete any non-read-only attributes set on this object.
    my @attrs = $self->_get_editable_attrs();
    my $attrs = $self->attrs();
    foreach my $attr (@attrs) {
        delete $attrs->{$attr};
    }

    # Reinitialize any non-read-only attributes.
    return $self->initialize();
}

#===============================================================================
# PROTECTED METHODS
#===============================================================================

#-------------------------------------------------------------------------------
# Object methods
#-------------------------------------------------------------------------------

#
# Overrides Class::Data::Inheritable::Translucent::_initialized().
#
sub _initialized {
    my $self = shift;
    return not exists $self->{__CDITB_initializing};
}

#
# Sets a private attribute in the object (sic; not the attributes hash returned
# by attrs()) indicating whether or not the object is currently initializing.
# This is intended to be set at the beginning of the constructor, immediately
# after creating the object body, and cleared just before returning from the
# constructor, just after initialization is complete.
#
sub _initializing {
    my($self, $flag) = @_;

    if ($flag) {
        $self->{__CDITB_initializing} = 1;
    }
    else {
        delete $self->{__CDITB_initializing};
    }
}

1;

__END__

#===============================================================================
# DOCUMENTATION
#===============================================================================

=head1 NAME

Class::Data::Inheritable::Translucent::Base - Add initialize/reset methods to Class::Data::Inheritable::Translucent

=head1 SYNOPSIS

See L<Class::Data::Inheritable::Translucent::Object> and
L<Class::Data::Inheritable::Translucent::Singleton>.

=head1 DESCRIPTION

This module is an abstract base class extending
L<Class::Data::Inheritable::Translucent> with common functionality
(initialize/reset methods) required by its
L<Class::Data::Inheritable::Translucent::Object> and
L<Class::Data::Inheritable::Translucent::Singleton> subclasses.

=head2 Object Methods

=over 4

=item C<initialize([ %args ])>

Sets the attributes of the invocant object from the given named parameter list
(hash) and returns 1.  If any given attributes are not object attributes or
translucent attributes of the class (or of any superclass) to which the
invocant object belongs, or if they are read-only and the object has already
been initialized, then they are ignored and initialize() will issue a warning
about them and return 0.

This method is called when constructing or cloning an object.  Subclasses may
override it to provide custom initialization, but it is not really intended to
be called directly.  In particular, if your object has read-only attributes then
initialize() will only be able to set them when it is called in the course of
constructing or cloning an object, but not if you call initialize() at a later
time, including when called via reset().

=item C<reset()>

Resets an object's non-read-only attributes to reveal their (translucent or
otherwise) class default values by deleting all such attributes from the object
and then calling initialize() with no arguments.

In the absence of any read-only attributes this has the effect of setting the
object to the same state as a new object constructed with a default constructor
call (i.e. new() with no arguments).

Read-only attributes, of course, cannot be reset so will be left unchanged.  If
initialize() has been overridden and is used to set any read-only attribute
values then it will emit a warning and return 0; otherwise it returns 1.

=back

=head1 DIAGNOSTICS

=head2 Warnings and Error Messages

This module may produce the following diagnostic messages.  They are classified
as follows (a la L<perldiag>):

    (W) A warning (optional).
    (F) A fatal error (trappable).
    (I) An internal error that you should never see (trappable).

=over 4

=item initialize() ignored the following read-only attributes: %s

(W) You passed some read-only attributes to a call to initialize() after the
object has been initialiazed (perhaps from via an override of initialize()
called from reset()).

=item initialize() ignored the following unknown attributes: %s

(W) You passed some attributes to initialize() (perhaps via new() or clone())
that are not object attributes or translucent attributes of the class (or of any
superclasses) to which the object being initialized (or constructed) belongs.

=item %s() is an object method, not a class method

(F) You tried to invoke the specified method on a class name, but it can only be
invoked on an object.

=back

=head1 SEE ALSO

L<Class::Data::Inheritable::Translucent::Object>,
L<Class::Data::Inheritable::Translucent::Singleton>,
L<Class::Data::Inheritable::Translucent>.

=head1 AUTHOR

Steve Hay E<lt>F<shay@cpan.org>E<gt>.

=head1 COPYRIGHT

Copyright (C) 2015 Steve Hay.  All rights reserved.

=head1 LICENCE

This module is free software; you can redistribute it and/or modify it under the
same terms as Perl itself, i.e. under the terms of either the GNU General Public
License or the Artistic License, as specified in the F<LICENCE> file.

=head1 VERSION

Version 2.00

=head1 DATE

TODO

=head1 HISTORY

See the F<Changes> file.

=cut

#===============================================================================


lib/Class/Data/Inheritable/Translucent/Object.pm
#===============================================================================
#
# lib/Class/Data/Inheritable/Translucent/Object.pm
#
# DESCRIPTION
#   Abstract base class extending Class::Data::Inheritable::Translucent with
#   constructor/clone methods with separated initialization.
#
# COPYRIGHT
#   Copyright (C) 2014-2015 Steve Hay.  All rights reserved.
#
# LICENCE
#   This module is free software; you can redistribute it and/or modify it under
#   the same terms as Perl itself, i.e. under the terms of either the GNU
#   General Public License or the Artistic License, as specified in the LICENCE
#   file.
#
#===============================================================================

package Class::Data::Inheritable::Translucent::Object;

use 5.008001;

use strict;
use warnings;

use parent qw(Class::Data::Inheritable::Translucent::Base);

use Carp qw(croak);

#===============================================================================
# CLASS INITIALIZATION
#===============================================================================

our($VERSION);

BEGIN {
    $VERSION = '2.00';
}

#===============================================================================
# PUBLIC METHODS
#===============================================================================

#-------------------------------------------------------------------------------
# Class methods
#-------------------------------------------------------------------------------

sub new {
    my($class, @args) = @_;
    croak("new() is a class method, not an object method") if ref $class;
    croak("'$class' is an abstract base class") if $class eq __PACKAGE__;

    my $self = bless {}, $class;

    $self->_initializing(1);

    # If there is an odd number of @args then the first should be an object to
    # copy into our new object. Otherwise leave the new object's attributes
    # uninitialized to provide translucency where appropriate.
    if (@args % 2) {
        my $source = shift @args;
        $self->copy($source);
    }

    # Initialize attributes with any (remaining) given arguments, overriding any
    # copied or translucent attributes in the process.
    $self->initialize(@args);

    $self->_initializing(0);

    return $self;
}

#-------------------------------------------------------------------------------
# Object methods
#-------------------------------------------------------------------------------

sub clone {
    my($self, %args) = @_;
    croak("clone() is an object method, not a class method") unless ref $self;

    my $class = ref $self;
    return $class->new($self, %args);
}

sub copy {
    my($self, $source) = @_;
    croak("copy() is an object method, not a class method") unless ref $self;

    if (not $source->isa(ref $self)) {
        my $class = ref $source;
        croak("Cannot copy attributes from instance of alien class '$class'");
    }

    # If the object has already been initialized then we must only copy
    # non-read-only attributes.
    my @attrs = $self->_get_editable_attrs();
    foreach my $attr (@attrs) {
        $self->$attr($source->$attr());
    }

    return 1;
}

1;

__END__

#===============================================================================
# DOCUMENTATION
#===============================================================================

=head1 NAME

Class::Data::Inheritable::Translucent::Object - Add constructor/clone methods to Class::Data::Inheritable::Translucent

=head1 SYNOPSIS

  package Bar;
  use parent qw(Class::Data::Inheritable::Translucent::Object);

  Bar->mk_object_accessor(attr => 0);

  # Class constructor
  my $obj1 = Bar->new();
  my $obj2 = Bar->new(attr => 1);
  my $obj3 = Bar->new($obj2);
  my $obj4 = Bar->new($obj2, attr => 2);

  # Object clone
  my $obj5 = $obj2->clone();
  my $obj6 = $obj2->clone(attr => 3);

  # Object copy
  my $obj7 = $obj2->copy();

  # Object reset
  $obj7->reset();

  package Bar2;
  use parent qw(Bar);

  my $_Count = 0;

  # Separated, overridable, object initialization
  sub initialize {
      my($self, %args) = @_;
      $_Count++;
      return $self->SUPER::initialize(%args);
  }

  sub DESTROY {
      $_Count--;
  }

=head1 DESCRIPTION

This module is an abstract base class extending
L<Class::Data::Inheritable::Translucent> with constructor/clone methods with
separated initialization.  The objects constructed are simple hash-based objects
compatible with the accessor methods installed by
Class::Data::Inheritable::Translucent.

Attributes may be initialized by the constructor, either explicitly, or by
copying another object, or via a separated initialize() method which can be
overridden in subclasses.  A copy() method is also provided to copy the
attributes of one object to another, and a reset() method can be used to restore
an object's non-read-only attributes to their default values (translucent or
otherwise).

Note that all of these methods, just like the accessor methods created by
L<Class::Data::Inheritable::Translucent>, only perform shallow copies of
attribute values.

=head2 Class Methods

=over 4

=item C<new([ $obj ] [, %args ])>

Constructs an instance of the invocant class and returns it.  The object will be
a simple hash-based object compatible with the accessor methods installed by
Class::Data::Inheritable::Translucent.

The new object can optionally be initialized either by copying the attributes of
a given object, or from a named parameter list (hash) of attributes, or both--in
which case the given object is cloned first and then any supplied attributes are
used to override the defaults and copied values.

=back

=head2 Object Methods

=over 4

=item C<clone($obj [, %args ])>

Clones the given object, constructing a new instance of the same class with the
same attribute values and returns it.

The new object can optionally have its default and copied attribute values
overridden by those in a named parameter list (hash) of attributes.

This has the same effect as $class->new($obj, %args), where $class is the class
to which $obj belongs.  In other words, this behaves like a copy constructor in
C++.

=item C<copy($obj)>

Sets the non-read-only attributes of the invocant object to the same values as
those of the given object, assuming that the given object is an instance of the
same class (or a subclass) as the invocant object (otherwise it throws an
exception), and returns 1.

Read-only attributes, of course, cannot be copied so will be left unchanged.

As noted earlier, but of particular relevance here, this only performs a shallow
copy of the attributes.  Therefore, the two objects will end up sharing some
data if any attribute values are references.

=item C<initialize([ %args ])>

Inherited from L<Class::Data::Inheritable::Translucent::Base>.

=item C<reset()>

Inherited from L<Class::Data::Inheritable::Translucent::Base>.

=back

=head1 DIAGNOSTICS

=head2 Warnings and Error Messages

This module may produce the following diagnostic messages.  They are classified
as follows (a la L<perldiag>):

    (W) A warning (optional).
    (F) A fatal error (trappable).
    (I) An internal error that you should never see (trappable).

=over 4

=item Cannot copy attributes from instance of alien class '%s'

(F) You tried to copy the attributes of an object belonging to the specified
class, which is not an instance of the same class (or of a superclass) as the
object you are trying to copy values to.

=item %s() is a class method, not an object method

(F) You tried to invoke the specified method on an object, but it can only be
invoked on a class name.

=item '%s' is an abstract base class

(F) You tried to construct an instance of the specified class, but it is an
abstract base class, i.e. it cannot be instantiated.

=item %s() is an object method, not a class method

(F) You tried to invoke the specified method on a class name, but it can only be
invoked on an object.

=back

=head1 SEE ALSO

L<Class::Data::Inheritable::Translucent::Singleton>,
L<Class::Data::Inheritable::Translucent>.

=head1 AUTHOR

Steve Hay E<lt>F<shay@cpan.org>E<gt>.

=head1 COPYRIGHT

Copyright (C) 2014-2015 Steve Hay.  All rights reserved.

=head1 LICENCE

This module is free software; you can redistribute it and/or modify it under the
same terms as Perl itself, i.e. under the terms of either the GNU General Public
License or the Artistic License, as specified in the F<LICENCE> file.

=head1 VERSION

Version 2.00

=head1 DATE

TODO

=head1 HISTORY

See the F<Changes> file.

=cut

#===============================================================================


Makefile.PL
#!perl
#===============================================================================
#
# Makefile.PL
#
# DESCRIPTION
#   Makefile creation script.
#
# COPYRIGHT
#   Copyright (C) 2014-2015 Steve Hay.  All rights reserved.
#
# LICENCE
#   This script is free software; you can redistribute it and/or modify it under
#   the same terms as Perl itself, i.e. under the terms of either the GNU
#   General Public License or the Artistic License, as specified in the LICENCE
#   file.
#
#===============================================================================

use 5.008001;

use strict;
use warnings;

use ExtUtils::MakeMaker 6.64;
use ExtUtils::MakeMaker qw(WriteMakefile);

#===============================================================================
# MAIN PROGRAM
#===============================================================================

MAIN: {
    WriteMakefile(
        NAME          => 'Class::Data::Inheritable::Translucent',
        ABSTRACT_FROM => 'lib/Class/Data/Inheritable/Translucent.pm',
        AUTHOR        => 'Ryan McGuigan, Steve Hay <shay@cpan.org>',
        LICENSE       => 'perl_5',
        VERSION_FROM  => 'lib/Class/Data/Inheritable/Translucent.pm',

        META_MERGE => {
            'meta-spec' => {
                version => 2
            },

            resources => {
                repository => {
                    type => 'git',
                    url  => 'https://github.com/steve-m-hay/Class-Data-Inheritable-Translucent.git'
                }
            },

            optional_features => {
                singleton => {
                    description => 'Class::Data::Inheritable::Translucent::Singleton',
                    prereqs => {
                        runtime => {
                            requires => {
                                'Apache::Singleton' => '0'
                            # or Class::Singleton
                            }
                        }
                    }
                },

                subname => {
                    description => 'Naming subroutines in caller() messages etc.',
                    prereqs => {
                        runtime => {
                            requires => {
                                'Sub::Name' => '0'
                            }
                        }
                    }
                },

                changestest => {
                    description => 'Changes testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::CPAN::Changes' => '0'
                            }
                        }
                    }
                },

                critictest => {
                    description => 'Perl::Critic testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Perl::Critic' => '0'
                            }
                        }
                    }
                },

                podtest => {
                    description => 'POD testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Pod'           => '1.14',
                                'Test::Pod::Coverage' => '1.04'
                            }
                        }
                    }
                },

                podcoveragetest => {
                    description => 'POD coverage testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Pod::Coverage' => '0.08'
                            }
                        }
                    }
                }
            }
        },

        MIN_PERL_VERSION => '5.008001',

        CONFIGURE_REQUIRES => {
            'ExtUtils::MakeMaker' => '6.64',
            'perl'                => '5.008001',
            'strict'              => '0',
            'warnings'            => '0'
        },

        TEST_REQUIRES => {
            'Test::More' => '0',
            'base'       => '0'
        },

        PREREQ_PM => {
            'Carp'       => '0',
            'Class::ISA' => '0',
            'constant'   => '0',
            'strict'     => '0',
            'warnings'   => '0'
        },

        dist => {
            PREOP   => 'find $(DISTVNAME) -type d -print|xargs chmod 0755 && ' .
                       'find $(DISTVNAME) -type f -print|xargs chmod 0644',
            TO_UNIX => 'find $(DISTVNAME) -type f -print|xargs dos2unix'
        }
    );
}

#===============================================================================
